tmt/steps/provision/__init__.py

```diff
@@ -146,3 +150,231 @@ def guest(self):
+        # override environment provided to execute().
+        environment.update(self.parent.plan.environment)
+        # Prepend with export and run as a separate command.
+        if not environment:
+            return ''
+        return 'export {}; '.format(
+            ' '.join(tmt.utils.shell_variables(environment)))
+
+    def ansible(self, playbook):
+        """ Prepare guest using ansible playbook """
+        playbook = self._ansible_playbook_path(playbook)
+        stdout, stderr = self.run(
+            f'stty cols {tmt.utils.OUTPUT_WIDTH}; ansible-playbook '
+            f'--ssh-common-args="{self._ssh_options(join=True)}" '
+            f'-e ansible_python_interpreter=auto'
+            f'{self._ansible_verbosity()} -i {self._ssh_guest()}, {playbook}')
+        self._ansible_summary(stdout)
+
+    def execute(self, command, **kwargs):
+        """
+        Execute command on the guest
+
+        command ....... string or list of command arguments
+        environment ... dictionary with environment variables
+        """
+
+        # Prepare the export of environment variables
+        environment = self._export_environment(kwargs.get('env', dict()))
+
+        # Change to given directory on guest if cwd provided
+        directory = kwargs.get('cwd', '')
+        if directory:
+            directory = f"cd '{directory}'; "
+
+        # Run in interactive mode if requested
+        interactive = ['-t'] if kwargs.get('interactive') else []
+
+        # Prepare command and run it
+        if isinstance(command, (list, tuple)):
+            command = ' '.join(command)
+        self.debug(f"Execute command '{command}' on guest '{self.guest}'.")
+        command = (
+            self._ssh_command() + interactive + [self._ssh_guest()] +
+            [f'{environment}{directory}{command}'])
+        return self.run(command, shell=False, **kwargs)
+
+    def push(self):
+        """ Push workdir to guest """
+        self.debug(f"Push workdir to guest '{self.guest}'.")
+        self.run(
+            f'rsync -Rrze "{self._ssh_command(join=True)}" --delete '
+            f'{self.parent.plan.workdir} {self._ssh_guest()}:/')
+
+    def pull(self):
+        """ Pull workdir from guest """
+        self.debug(f"Pull workdir from guest '{self.guest}'.")
+        self.run(
+            f'rsync -Rrze "{self._ssh_command(join=True)}" '
+            f'{self._ssh_guest()}:{self.parent.plan.workdir} /')
+
+    def stop(self):
+        """
+        Stop the guest
+
+        Shut down a running guest instance so that it does not consume
+        any memory or cpu resources. If needed, perform any actions
+        necessary to store the instance status to disk.
+        """
+        self.debug(f"Doing nothing to stop guest '{self.guest}'.")
+
+    def remove(self):
+        """
+        Remove the guest
+
+        Completely remove all guest instance data so that it does not
+        consume any disk resources.
```

Comment:
What about artefacts, like downloaded base images? In case of vagrant, there are `base-boxes` and `images` for containers. Those are preserved, right?

_Theoretically:_ that would be `clean()`, right?

Reply:
Yes, agreed, we can add something like `clean()` in the future.